File: src/com/minhle/practiceleetcode/SubArraySumDivisibleByK.cpp
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        vector<int> prefix(nums.size(), 0);
        prefix.at(0) = ((nums.at(0) % k) + k) % k;
        for (int i = 1; i < nums.size(); ++i)
        {
            prefix.at(i) = (((prefix.at(i - 1) + nums.at(i)) % k) + k) % k; 
        }
        int res = 0;
        map<int, int> frequency;
        frequency[0] = 1;
        for (int sum : prefix)
        {
            if (frequency.find(sum) == frequency.end())
            {
                frequency[sum] = 0;
            }
            ++frequency[sum];
        }
        for (const auto &pair : frequency)
        {
            if (pair.second > 1)
            {
                res += (pair.second * (pair.second - 1)) / 2;
            }
        }
        return res;
    }
};

File: src/com/minhle/practiceleetcode/SubarraySumDivisibleByK.java
package com.minhle.practiceleetcode;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class SubarraySumDivisibleByK {
    public int subarraysDivByK(int[] nums, int k) {
        int[] cumulativeSum = new int[nums.length + 1];
        for (int i = 1; i < cumulativeSum.length; i++) {
            cumulativeSum[i] = cumulativeSum[i - 1] + nums[i - 1];
        }
        int count = 0;
        for (int i = 0; i < cumulativeSum.length - 1; i++) {
            for (int j = i + 1; j < cumulativeSum.length; j++) {
                int sum = cumulativeSum[j] - cumulativeSum[i];
                if (sum % k == 0){
                    count++;
                }
            }
        }
        return count;
    }

    @Test
    void test1() {
        int[] nums = {4,5,0,-2,-3,1};
        int k = 5;
        Assertions.assertEquals(7, subarraysDivByK(nums, k));
    }
}
